File: Part1_강의자료/Ch08_투포인터/문제별코드/12891_DNA비밀번호/12891.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>
using namespace std;

bool is_valid_sequence(int minimum_base_count[4], int current_base_count[4])
{
    for (int i = 0; i < 4; i++)
        if (current_base_count[i] < minimum_base_count[i])
            return false;
    return true;
}

int main()
{
    int S, P, minimum_base_count[4];
    string sequence;
    cin >> S >> P >> sequence;
    for (int i = 0; i < 4; i++)
        cin >> minimum_base_count[i];

    map<char, int> dic;
    dic['A'] = 0;
    dic['C'] = 1;
    dic['G'] = 2;
    dic['T'] = 3;

    int current_base_count[4] = { 0 };
    for (int i = 0; i < P; i++)
        current_base_count[dic[sequence[i]]]++;

    int ans = is_valid_sequence(minimum_base_count, current_base_count);
    for (int i = 1; i <= S - P; i++)
    {
        current_base_count[dic[sequence[i - 1]]]--;
        current_base_count[dic[sequence[i + P - 1]]]++;
        ans += is_valid_sequence(minimum_base_count, current_base_count);
    }
    cout << ans << '\n';
    return 0;
}

File: Part1_강의자료/Ch08_투포인터/문제별코드/12891_DNA비밀번호/12891.py
S, P = map(int, input().split())
sequence = input()
minimum_base_count = list(map(int, input().split()))
dic = {'A': 0, 'C': 1, 'G': 2, 'T': 3}

current_base_count = [0] * 4
for ch in sequence[:P]:
    current_base_count[dic[ch]] += 1
    
def is_valid_sequence():
    for i in range(4):
        if current_base_count[i] < minimum_base_count[i]:
            return 0
    return 1

ans = is_valid_sequence()
for i in range(1, S - P + 1):
    current_base_count[dic[sequence[i - 1]]] -= 1
    current_base_count[dic[sequence[i + P - 1]]] += 1
    ans += is_valid_sequence()

print(ans)
